lms/static/scripts/frontend_apps/components/dashboard/SyncGradesButton.tsx

```diff
@@ -0,0 +1,120 @@
+import {
+  Button,
+  LeaveIcon,
+  SpinnerCircleIcon,
+} from '@hypothesis/frontend-shared';
+import { useCallback, useMemo, useState } from 'preact/hooks';
+import { useParams } from 'wouter-preact';
+
+import type { GradingSync } from '../../api-types';
+import { useConfig } from '../../config';
+import { apiCall, usePolledAPIFetch } from '../../utils/api';
+import type { QueryParams } from '../../utils/url';
+import { replaceURLParams } from '../../utils/url';
+
+export type SyncGradesButtonProps = {
+  /**
+   * List of students and their grades, which should be synced when the button
+   * is clicked.
+   * Passing `undefined` means students are not yet known and/or being loaded.
+   */
+  studentsToSync?: Array<{ h_userid: string; grade: number }>;
+};
+
+export default function SyncGradesButton({
+  studentsToSync,
+}: SyncGradesButtonProps) {
+  const { assignmentId } = useParams<{ assignmentId: string }>();
+  const { dashboard, api } = useConfig(['dashboard', 'api']);
+  const { routes } = dashboard;
+
+  const syncUrl = useMemo(
+    () =>
+      replaceURLParams(routes.assignment_grades_sync, {
+        assignment_id: assignmentId,
+      }),
+    [assignmentId, routes.assignment_grades_sync],
+  );
+  const [lastSyncParams, setLastSyncParams] = useState<QueryParams>({});
+  const lastSync = usePolledAPIFetch<GradingSync>({
+    path: syncUrl,
+    params: lastSyncParams,
+    // Keep polling as long as sync is in progress
+    shouldRefresh: result =>
+      !!result.data &&
+      ['scheduled', 'in_progress'].includes(result.data.status),
+  });
+
+  const buttonContent = useMemo(() => {
+    if (!studentsToSync || (lastSync.isLoading && !lastSync.data)) {
+      return 'Loading...';
+    }
+
+    if (
+      lastSync.data &&
+      ['scheduled', 'in_progress'].includes(lastSync.data.status)
+    ) {
+      return (
+        <>
+          Syncing grades
+          <SpinnerCircleIcon className="ml-1.5" />
+        </>
+      );
+    }
+
+    // TODO Maybe these two should be represented differently
+    if (lastSync.error || lastSync.data?.status === 'failed') {
+      return (
+        <>
+          Error syncing. Click to retry
+          <LeaveIcon />
+        </>
+      );
+    }
+
+    if (studentsToSync.length > 0) {
+      return `Sync ${studentsToSync.length} students`;
+    }
+
+    return 'Grades synced';
+  }, [studentsToSync, lastSync.isLoading, lastSync.data, lastSync.error]);
+
+  const buttonDisabled = useMemo(
```

Comment:
The current logic of this `useMemo` is so cheap the `useMemo` will likely add overhead.